### PR TITLE
binding/c: update python scripts

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Split from PR #4941 

This PR updates the python binding scripts as a per-requisite for the rest of the API generation. 

The actual impact of this PR is when we enable the generation of rest of API functions. We can always make necessary patches then.

Added --

Ability and option to directly load `apis.json` as from the standard repository build process. If we import the `apis.json` to `src/binding/` folder, it will load apis.json instead of `mpi_standards_api.txt`.

Use `re.IGNORECASE` instead of `re.I`

Better temporary variable names.

Added `.docnotes` directive for additional docnotes

Added `.errorcodes` directives for additional error codes.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
@shintaro-iwasaki I think you already reviewed most of the commits in this PR. Could review it again or approve it while I make sure the tests are good?

**Review by shintaro-iwasaki — APPROVED** [review]
Except for "binding/c: update python scripts" which is too large to review, this PR looks okay to me.

**Comment by hzhou** [comment]
test:mpich/ch4/most

File: maint/gen_binding_c.py
Comment (raffenet) [inline maint/gen_binding_c.py:8]: it is not possible for me to do any meaningful review of this commit. there are 1500 lines of change. i don't know what to do with it besides look at the output.

Comment (hzhou) [reply maint/gen_binding_c.py:8]: I know. Unless there is something you particularly notice, such as styles or obvious strangeness, I think we can merge them. They should not have impact to the current code other than the pt2pt chapter that already got merged. We'll review the output in the following chap-by-chap PRs and if there is errors, we can always add patches then.

File: maint/gen_binding_c.py
Comment (raffenet) [inline maint/gen_binding_c.py:31]: IMO, regex is overkill for simple substring matches. This can be done with `in` or `.find()`, with the benefit that it should perform better.

Comment (hzhou) [reply maint/gen_binding_c.py:31]: I can change that.

Regex is fast enough, so I wouldn't worry about performance optimization unless there is evidence and justification. One of the drawback is regex tend to become too complex. That can be fixed by restructure the code design to keep the regex simple.

The benefit of regex vs direct string comparison or `in` or `find` is flexibility. For example, it can easily accommodate extra spaces, multiple spelling variations, flexible cases, etc. without significantly change the code appearance or structure. The alternative of using a mix of `==`, `in`, `find`, and regex, imo, can introduce unnecessary visual code complexity.

So I hope I convinced you to overlook this detail, but if you insist, I'll change it (and potentially many other places).

File: maint/local_python/binding_c.py
Comment (raffenet) [inline maint/local_python/binding_c.py:203]: what is `pj`?

File: maint/local_python/binding_c.py
Comment (raffenet) [inline maint/local_python/binding_c.py:725]: Single letter variable names are really hard to understand. We should fix all the instances in these scripts.

File: maint/local_python/binding_c.py
Comment (raffenet) [inline maint/local_python/binding_c.py:724]: We should use the long form `re.IGNORECASE` to make intent clearer.

File: src/include/mpir_impl_tmp.h
Comment (raffenet) [inline src/include/mpir_impl_tmp.h:6]: I feel we can separately review the generation of `mpir_impl.h` after the MPI/PMPI work is merged. Or is there some reason this needs to be done together?

File: src/mpi/errhan/errutil.c
Comment (raffenet) [inline src/mpi/errhan/errutil.c:1696]: Is this needed? Or just an improvement to the error output? I can't see why returning the predefined integer values is wrong.

File: maint/local_python/binding_c.py
Comment (hzhou) [reply maint/local_python/binding_c.py:203]: A temporary variable name for `p` since `p` is already used.  Will `p_j` reads better?

File: maint/local_python/binding_c.py
Comment (hzhou) [reply maint/local_python/binding_c.py:725]: There are only a couple --
* `G` for globals. It is more of a marker since Python has no facility to distinguish global or local variables and it has no scopes. 
*  `p` for parameter, `param` better?
*  `a` used for parsing arguments and lists, it is scoped within a few lines. I often find it more confusing for very local temporary variables to have longer names. The latter the reader will tempted to search for meanings elsewhere while there isn't any. It is similar to `i`, and `j`.

File: maint/local_python/binding_c.py
Comment (hzhou) [reply maint/local_python/binding_c.py:724]: Okay

File: src/include/mpir_impl_tmp.h
Comment (hzhou) [reply src/include/mpir_impl_tmp.h:6]: If we don't generate `mpir_impl.h`, then I have to manually declare the prototypes, as I did with `pt2pt` and during refactoring of `attr` and `errorhan`. That is the reason `mpir_impl.h` was created in the first place. We have to manually check which function needs the prototypes since many but not all in every chapter has slightly changed functions names (mostly with or without `_impl` suffix).  This is a very tedious work that will get deleted in the end. So I am hoping that we can avoid it.

File: src/mpi/errhan/errutil.c
Comment (hzhou) [reply src/mpi/errhan/errutil.c:1696]: It is for consistency. I don't see why the KEYVAL handle object is any different from other handle objects. The integer value wouldn't make any sense to the user -- at least we should print it as hexadecimals.

File: maint/local_python/binding_c.py
Comment (hzhou) [reply maint/local_python/binding_c.py:724]: mark as fixed (once pushed)

File: maint/mpi_standard_api.txt
Comment (raffenet) [inline maint/mpi_standard_api.txt:153]: has the script used to scrape the standard data been committed to the repo?

Comment (hzhou) [reply maint/mpi_standard_api.txt:153]: No. 

The mpi standards -- at least the API functions -- are not expected to change often. I think it is simpler to directly maintain `mpi_standard_api.txt` directly (even manually), rather than having to maintain one more script and add a dependency on the private mpi-standards repository.

Comment (raffenet) [reply maint/mpi_standard_api.txt:153]: I don't think this is a good idea. There is a hidden step to this that only you currently understand.

Comment (hzhou) [reply maint/mpi_standard_api.txt:153]: If I say I manually transcribed the api interface into this config file, would that make any difference?

The script I use is in Perl and it is for one-time purpose that I don't intend to maintain (other than hack). This config file will always need be directly maintained anyway since we want to review explicit changes from this point on. If we get initial transcription correct, I expect it is not a big task in the future to manually update this file.

It is one of the intentions to decouple this config file from the python-framework used in mpi-standards. That python layer are subject to change with no apparent reasons from outside. For example, they may change the way how the categorize parameters, variable names, particular attributes syntaxes, etc.. I don't think we really want to follow those changes. Any changes to `mpi_standard_api.txt` in the future should be treated as an update from standards from user point of view.

Comment (hzhou) [reply maint/mpi_standard_api.txt:153]: Besides, if we want a *correct* scraper, we'll need the scraper to be able to parse full Python syntax. That is way too complex and unnecessary. But without the *correctness*, the script is considered hacky and thus I am unwilling to maintain.

Comment (raffenet) [reply maint/mpi_standard_api.txt:153]: > It is one of the intentions to decouple this config file from the python-framework used in mpi-standards.

I disagree with this intent. The source data in the MPI standard is json. Python includes a json parser in the standard library. I don't think this is a big ask.

Comment (hzhou) [reply maint/mpi_standard_api.txt:153]: It is not json. It is full python code:
```
\begin{mpi-binding}
    function_name("MPI_Send")

    parameter(
        "buf", "BUFFER", desc="initial address of send buffer", constant=True
    )
    parameter(
        "count",
        "XFER_NUM_ELEM_NNI_SMALL",
        desc="number of elements in send buffer",
    )
    parameter(
        "datatype", "DATATYPE", desc="datatype of each send buffer element"
    )
    parameter("dest", "RANK", desc="rank of destination")
    parameter("tag", "TAG", desc="message tag")
    parameter("comm", "COMMUNICATOR")
\end{mpi-binding}
```

To do it "correctly", not only we need parse python syntax, we also need import all their python stack. Probably that the only way is to directly import their code, then we are having a hard dependency on it.

Comment (raffenet) [reply maint/mpi_standard_api.txt:153]: If you build the document, there is an `apis.json` file generated.

Comment (hzhou) [reply maint/mpi_standard_api.txt:153]: Also their python code is focused on generating 'LaTeX', so importing it still doesn't reduce the complexity much on our part.

Comment (raffenet) [reply maint/mpi_standard_api.txt:153]: I'm not talking about importing their Python code. Please don't misunderstand me.

Comment (hzhou) [reply maint/mpi_standard_api.txt:153]: I didn't realize the generated `apis.json`. We can directly use the json file -- i.e. directly add `apis.json` to our repository. Is that more preferable. The apiis.json doesn't have language bindings (parameter types), so that still need be separately scraped.

File: maint/local_python/binding_c.py
Comment (hzhou) [reply maint/local_python/binding_c.py]: changed into `temp_p`

File: maint/mpi_standard_api.txt
Comment (raffenet) [reply maint/mpi_standard_api.txt]: It might be worth engaging the standard maintainers to see if parameter types might be something they can add to the json, or if some other solution exists.

I don't want to block others, so I think we can move forward with this. The API source information is a big piece of technical debt that I think we should address in the long-term.

**Comment by hzhou** [comment]
@raffenet I added two additional commits. The last one loads `apis.json` directly. `apis.json` is 56,254 lines long 😱 (vs. `mpi_standard_api.txt` 3,432 lines) 😄 Anyway, could check if the new version is preferrable?

**Comment by hzhou** [comment]
@raffenet I removed the `apis.json` but retained the ability to load `apis.json`, so we always have a fallback option and it is good if we want to verify the correctness of `mpi_standards_api.txt`.

I addressed your comment on `re.IGNORECASE` and better temporary variable name.

I added `.docnotes` and `.errorcodes` directives to reflect the review from @shintaro-iwasaki. 

Do you see any other issues I need to address for this PR?

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Review by raffenet — APPROVED** [review]
We had a lot of offline discussion. Short-term, I think this is fine to merge and unblock everyone else.

Long-term, I hope we can better leverage what the MPI standard document provides and avoid the custom data format and parser API information. I have reached out to the people maintaining this info in the standard, and `apis.json` is currently the best source of API data. However, more work is planned to better facilitate code generation and they are looking for input from folks like us. I think it would be mutually beneficial if we can contribute to their efforts.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
@raffenet I added two more addition trivial commits to this PR. If there is no further comments, I will merge after tests clear. Otherwise, let me know.

**Review by raffenet — APPROVED** [review]

**Pull request merged**
